Guided Transfer: Complete purchase notice

jordwest opened the pull request:
~~**Code Review: Please only review d74b44a. Prior commits belong to a refactor (#7347) which this is rebased on top of.**~~ #7347 is now merged into master

This adds a notice to the export page of site settings prompting the user to complete a Guided Transfer purchase.

This will appear when the user has entered host details, but hasn't yet completed checkout.

![screen shot 2016-08-01 at 6 15 56 pm](https://cloud.githubusercontent.com/assets/416133/17287611/09aa9030-5814-11e6-8004-a123e2636997.png)
### How to test

~~Currently this is disabled until we have the required data in the Redux store (#6471). In order to test, manually set `isGuidedTransferAwaitingPurchase: true` in `client/my-sites/exporter/index.jsx` on L45~~
1. Go to **My Sites > Settings > Export**
2. Open the Redux devtools. Down the bottom, click **Dispatcher**.
3. Paste the following action (replacing YOUR_SITE_ID_HERE) then press Dispatch:
   
   ``` js
   {
     type: 'GUIDED_TRANSFER_STATUS_RECEIVE',
     siteId: YOUR_SITE_ID_HERE,
     guidedTransferStatus: {
       issues: [],
       upgrade_purchased: false,
       host_details_entered: true,
     }
   }
   ```
4. You should immediately see the orange notice as above.
5. Paste the following into the dispatcher
   
   ``` js
   {
     type: 'GUIDED_TRANSFER_STATUS_RECEIVE',
     siteId: YOUR_SITE_ID_HERE,
     guidedTransferStatus: {
       issues: [],
       upgrade_purchased: true,
       host_details_entered: true,
     }
   }
   ```
6. You should now see the in progress screen, and the Guided Transfer card should disappear:

![screen shot 2016-08-15 at 4 27 07 pm](https://cloud.githubusercontent.com/assets/416133/17657281/31093f54-6305-11e6-83f2-f5f031225f93.png)

cc @dllh @mrheston @omarjackman 

Test live: https://calypso.live/?branch=add/guided-transfer/complete-purchase-notice

jordwest committed:
Guided Transfer: Complete purchase notice

This adds a notice to the export page of site settings prompting the user to complete a Guided Transfer purchase. This should
appear when the user has entered host details, but hasn't yet completed checkout.

Change 'Purchase' button text to 'Continue'

Connects notice to Redux

extra JSDoc

Connect in progress card to Redux

folletto commented:
Design wise:

1. Styling is ok. The warning message is maybe a bit too much, but I think it's worth it.
2. What happens if the user tries to initiate a new guided transfer while the other it's still pending?
3. What are the conditions that trigger it, and what are the conditions that dismiss it?
4. Can a user dismiss it manually?

jordwest commented:
Thanks for taking a look!

> What happens if the user tries to initiate a new guided transfer while the other it's still pending?

There can only be a single Guided Transfer per site; attempting to purchase or initiate another guided transfer would fail, although the UI won't allow the user to attempt it. Once the GT is purchased and necessary details are entered, the GT card will be replaced with an "in progress" card (see #6396).

> What are the conditions that trigger it

Clicking the 'Purchase a Guided Transfer' button will take the user through a series of steps to gather information about the destination host. Once they've saved that information, they'll be taken through the checkout and purchase the guided transfer. In those cases, the user will never see this notice. However, in some cases they might close the tab before completing payment, and that's the case that this notice takes care of.

> What are the conditions that dismiss it?

Completing the purchase will replace this notice with the in progress card (#6396). If the user never makes the purchase, the saved destination host details are purged after a certain period of time and this notice will disappear.

> Can a user dismiss it manually?

No, presumably if the user has entered destination host details there's a good chance they want to continue.

folletto commented:
Cool, thanks! 👍

mtias commented:
Not entirely sure of the whole flow, but "Purchase" doesn't seem like the best word to describe the action if the user has already started the process. Perhaps "Proceed with purchase" or "Continue"?

jordwest commented:
@mtias: That makes sense, updated:

![screen shot 2016-08-02 at 5 28 15 pm](https://cloud.githubusercontent.com/assets/416133/17320359/89896d7c-58d6-11e6-8614-5c55eb8f2032.png)

omarjackman commented on client/my-sites/exporter/exporter.jsx:
```diff
@@ -8,6 +8,7 @@ import React, { PropTypes } from 'react';
  */
 import FoldableCard from 'components/foldable-card';
 import GuidedTransferInProgress from './guided-transfer-in-progress';
+import GuidedTransferAwaitingPurchase from '../guided-transfer/awaiting-purchase-notice';
```
Shouldn't this be `'my-sites/guided-transfer/awaiting-purchase-notice'`

jordwest replied:
Absolute paths FTW. Fixed in b2b675f

jordwest commented:
These changes are going to conflict with #7347 - I'll rebase on that PR before putting this up for review again

jordwest commented:
This is now ready for review again, but because it includes the changes in #7347, I've squashed the current PR's changes into d74b44a. Please review only that commit in this PR.

omarjackman commented:
@jordwest The code looks good to me and I tested the way you outlined but I noticed that after I dispatched the event, clicked on "Continue", and then clicked the back arrow, the notice wasn't there anymore. Is that just because the devtools dispatched the event but the status wasn't actually added to the store?

jordwest commented:
@omarjackman Yep, the dispatch only overrides the local store, so when returning to the export page it fetches the real status from the server again.

omarjackman commented:
@jordwest I just tested this again and I couldn't get either notices to display

omarjackman commented:
@jordwest I got it to work but I needed to clear my local storage first

omarjackman commented:
@jordwest I think the changes you've made look good. I was a bit worried that I had to clear local storage for it to begin working but that could just be a result of using the redux devtools. If you've tested this to the point you feel its solid then LGTM.

jordwest commented:
Hmm that's weird, it's persisted in the Redux store (which generally uses IndexedDB) so localStorage shouldn't have any effect on it. I have had some issues where the Redux devtools didn't actually do the dispatch, and restarting the browser fixed it, so that could be related. Did you notice if the dispatched action appeared in the action replay history in devtools?

 I'll do some more testing today before merging.

jordwest commented:
Did more testing on this and couldn't reproduce, was likely just a quirk in the devtools. Going ahead for merge

jordwest committed:
Don't load GT data for all environments

Disables querying of GT data for environments where the
feature flag is disabled.

jordwest commented:
Just made a small tweak to ensure we don't load GT data for environments where the feature flag isn't yet enabled
